src/main/flight/altitude.c

```diff
@@ -203,122 +210,104 @@ int32_t calculateAltHoldThrottleAdjustment(int32_t vel_tmp, float accZ_tmp, floa
+    const uint32_t dTime = currentTimeUs - previousTimeUs;
+    if (dTime < BARO_UPDATE_FREQUENCY_40HZ) {
         return;
-
+    }
     previousTimeUs = currentTimeUs;
 
+    static float vel = 0.0f;
+    static float accAlt = 0.0f;
+
+    int32_t baroAlt = 0;
 #ifdef BARO
-    if (!isBaroCalibrationComplete()) {
-        performBaroCalibrationCycle();
-        vel = 0;
-        accAlt = 0;
+    if (sensors(SENSOR_BARO)) {
+        if (!isBaroCalibrationComplete()) {
+            performBaroCalibrationCycle();
+            vel = 0;
+            accAlt = 0;
+        } else {
+            baroAlt = baroCalculateAltitude();
+            estimatedAltitude = baroAlt;
+        }
     }
-
-    baro.BaroAlt = baroCalculateAltitude();
-#else
-    baro.BaroAlt = 0;
 #endif
 
 #ifdef SONAR
-    sonarAlt = sonarRead();
-    sonarAlt = sonarCalculateAltitude(sonarAlt, getCosTiltAngle());
-
-    if (sonarAlt > 0 && sonarAlt < sonarCfAltCm) {
-        // just use the SONAR
-        baroAlt_offset = baro.BaroAlt - sonarAlt;
-        baro.BaroAlt = sonarAlt;
-    } else {
-        baro.BaroAlt -= baroAlt_offset;
-        if (sonarAlt > 0  && sonarAlt <= sonarMaxAltWithTiltCm) {
+    if (sensors(SENSOR_SONAR)) {
+        int32_t sonarAlt = sonarRead();
+        sonarAlt = sonarCalculateAltitude(sonarAlt, getCosTiltAngle());
+        if (sonarAlt > 0 && sonarAlt >= sonarCfAltCm && sonarAlt <= sonarMaxAltWithTiltCm) {
             // SONAR in range, so use complementary filter
-            sonarTransition = (float)(sonarMaxAltWithTiltCm - sonarAlt) / (sonarMaxAltWithTiltCm - sonarCfAltCm);
-            baro.BaroAlt = sonarAlt * sonarTransition + baro.BaroAlt * (1.0f - sonarTransition);
+            float sonarTransition = (float)(sonarMaxAltWithTiltCm - sonarAlt) / (sonarMaxAltWithTiltCm - sonarCfAltCm);
+            sonarAlt = sonarAlt * sonarTransition + baroAlt * (1.0f - sonarTransition);
+            estimatedAltitude = sonarAlt;
         }
     }
 #endif
 
-    dt = accTimeSum * 1e-6f; // delta acc reading time in seconds
+    float accZ_tmp = 0;
+#ifdef ACC
+    if (sensors(SENSOR_ACC)) {
+        const float dt = accTimeSum * 1e-6f; // delta acc reading time in seconds
 
-    // Integrator - velocity, cm/sec
-    if (accSumCount) {
-        accZ_tmp = (float)accSum[2] / (float)accSumCount;
-    } else {
-        accZ_tmp = 0;
-    }
-    vel_acc = accZ_tmp * accVelScale * (float)accTimeSum;
-
-    // Integrator - Altitude in cm
-    accAlt += (vel_acc * 0.5f) * dt + vel * dt;                                                                 // integrate velocity to get distance (x= a/2 * t^2)
-    accAlt = accAlt * barometerConfig()->baro_cf_alt + (float)baro.BaroAlt * (1.0f - barometerConfig()->baro_cf_alt);    // complementary filter for altitude estimation (baro & acc)
-    vel += vel_acc;
+        // Integrator - velocity, cm/sec
+        if (accSumCount) {
+            accZ_tmp = (float)accSum[2] / (float)accSumCount;
```

Comment:
It's enough to cast either numerator or denominator to `float`, other side will get an implicit floating conversion.